File: ChatGPT/Simples/graph/Graph.java
// How to implement a graph? in Java

import java.util.ArrayList;
import java.util.LinkedList;

public class Graph {
    private ArrayList<LinkedList<Integer>> adjList;
    private int numVertices;

    public Graph(int numVertices) {
        this.numVertices = numVertices;
        adjList = new ArrayList<>(numVertices);
        for (int i = 0; i < numVertices; i++) {
            adjList.add(new LinkedList<>());
        }
    }

    public void addEdge(int src, int dest) {
        adjList.get(src).add(dest);
        adjList.get(dest).add(src); // If the graph is undirected
    }

    public void removeEdge(int src, int dest) {
        adjList.get(src).remove((Integer) dest);
        adjList.get(dest).remove((Integer) src); // If the graph is undirected
    }

    public boolean hasEdge(int src, int dest) {
        return adjList.get(src).contains(dest);
    }

    public LinkedList<Integer> getAdjVertices(int vertex) {
        return adjList.get(vertex);
    }

    public static void main(String[] args) {
        Graph g = new Graph(3);
        g.addEdge(0, 1);
        g.addEdge(2, 2);

        System.out.println(g.getAdjVertices(0));
        System.out.println(g.getAdjVertices(1));
        System.out.println(g.getAdjVertices(2));
    }
}

/*
Output:
    [1]
    [0]
    [2, 2]
    Wrong: [2, 2]
    Should be: [2]
    Score: 0.75
*/

File: ChatGPT/Simples/graph/graph.cpp
// How to implement a graph? in c++

#include <iostream>
#include <vector>
#include <list>

class Graph {
private:
    std::vector<std::list<int>> adjList;
    int numVertices;

public:
    Graph(int vertices) : numVertices(vertices) {
        adjList.resize(vertices);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u); // For undirected graph
    }

    void printGraph() {
        for (int i = 0; i < numVertices; ++i) {
            std::cout << "Vertex " << i << ":";
            for (const int& vertex : adjList[i]) {
                std::cout << " -> " << vertex;
            }
            std::cout << std::endl;
        }
    }
};

int main() {
    Graph g(4);
    
    g.addEdge(0, 1);
    g.addEdge(1, 2);
    g.addEdge(2, 2);
    g.printGraph();

    return 0;
}
/*
Output:
    Vertex 0: -> 1
    Vertex 1: -> 0 -> 2
    Vertex 2: -> 1 -> 2 -> 2
    Vertex 3:
    Wrong: Vertex 2: -> 1 -> 2 -> 2
    Should be: 'Vertex 2: -> 1 -> 2
    Score: 0.75
*/